### PR TITLE
[Draft] [FP8] CUTLASS FP8 matrix multiply

### PR DESCRIPTION
**Pull request by pcmoritz**
This is a relatively clean port of the CUTLASS FP8 code from https://github.com/NVIDIA/cutlass/blob/033d9efd2db0bbbcf3b3b0650acde6c472f3948e/examples/54_hopper_fp8_warp_specialized_gemm/54_hopper_fp8_warp_specialized_gemm.cu

No scaling yet and some of the numerics might not be right. I tried to focus on improving the small batch size matrix multiplication performance for bs x 4096 x 4096 multiplications.

With @mgoin 's benchmark script from https://github.com/vllm-project/vllm/pull/4118#issuecomment-2061935623, https://github.com/mgoin/torch-fp8/blob/main/benchmark_scaled_mm.py with the diff in the appendix, I'm getting the following results:

**This is comparing this code (FP8) and the scaled_mm code, both with static scaling, with FP16**

```
M, N, K, Cosine similarity, FP16 time, FP8 time, FP8 scaled_mm, Speedup, Speedup scaled_mm
1, 4096, 4096, 0.9995, 0.1176, 0.1511, 0.2035, 0.778, 0.578
2, 4096, 4096, 0.9995, 0.1173, 0.1374, 0.1962, 0.853, 0.598
4, 4096, 4096, 0.9990, 0.1145, 0.1379, 0.1876, 0.830, 0.610
8, 4096, 4096, 0.9990, 0.1171, 0.1329, 0.1875, 0.881, 0.624
16, 4096, 4096, 0.9990, 0.1204, 0.1302, 0.1887, 0.925, 0.638
32, 4096, 4096, 0.9995, 0.1209, 0.1219, 0.1819, 0.992, 0.665
64, 4096, 4096, 0.9995, 0.1464, 0.1174, 0.1804, 1.247, 0.811
128, 4096, 4096, 0.9995, 0.1666, 0.1190, 0.1803, 1.399, 0.924
256, 4096, 4096, 0.0000, 0.2359, 0.2165, 0.1806, 1.090, 1.307
```

The TL;DR: Some improvements, but not nearly there yet.

appendix:
```diff
diff benchmark_scaled_mm2.py benchmark_scaled_mm.py
2a3
> from vllm import _C
14a16
>     y = torch.empty_like(x)
20a23,24
>     workspace = torch.zeros(32 * 1024 * 1024, dtype=torch.uint8, device='cuda')
> 
24c28,29
<         y, _ = torch._scaled_mm(x_fp8, w_fp8, out_dtype=dtype, scale_a=x_inv_s, scale_b=w_inv_s)
---
>         # y, _ = torch._scaled_mm(x_fp8, w_fp8, out_dtype=dtype, scale_a=x_inv_s, scale_b=w_inv_s)
>         _C.ops.fp8_scaled_gemm(y, x_fp8, w_fp8, workspace)
32c37,38
<         y, _ = torch._scaled_mm(x_fp8, w_fp8, out_dtype=dtype, scale_a=x_inv_s, scale_b=w_inv_s)
---
>         _C.ops.fp8_scaled_gemm(y, x_fp8, w_fp8, workspace)
>         # y, _ = torch._scaled_mm(x_fp8, w_fp8, out_dtype=dtype, scale_a=x_inv_s, scale_b=w_inv_s)
41c47
<         x_fp8, x_inv_s = to_float8(x, dtype=qdtype)
---
>         # x_fp8, x_inv_s = to_float8(x, dtype=qdtype)
75c81
<     print("M, N, K, Cosine similarity, FP16 time, FP8 time, FP8 dynamic quant time, Speedup, Speedup dynamic quant")
---
>     print("M, N, K, Cosine similarity, FP16 time, FP8 time, FP8 scaled_mm, Speedup, Speedup scaled_mm")
```

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by hmellor** [comment]
Hi @pcmoritz, is @cyang49 right that this CUTLASS code has now already been ported?
